File: src/main/java/hm/binkley/rover/CStyleMain.java
package hm.binkley.rover;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import static java.lang.System.in;
import static java.lang.System.out;
import static java.util.stream.Collectors.toList;

/**
 * {@code CStyleMain} is a very "C"-like approach to the Rover problem.
 *
 * @author <a href="mailto:binkley@alumni.rice.edu">B. K. Oxley (binkley)</a>
 * @see <a href="http://arunrocks.com/mars-rover-in-python-and-haskell/">Mars
 * Rover in Python and Haskell</a>
 */
public final class CStyleMain {
    public static void main(final String... args)
            throws IOException {
        final var lines = new BufferedReader(new InputStreamReader(in))
                .lines()
                .collect(toList());

        if (0 == lines.size() % 2)
            throw new IOException("Malformed input");

        // TODO: Skip grid size - what to do with it?
        for (int i = 1, l = lines.size(); i < l; ++i) {
            final String co = lines.get(i);
            final String[] coords = co.split(" ", 3);
            int x = Integer.parseInt(coords[0]);
            int y = Integer.parseInt(coords[1]);
            int d = "ENWS".indexOf(coords[2].charAt(0));

            final String mo = lines.get(++i);
            for (int j = 0, k = mo.length(); j < k; ++j) {
                switch (mo.charAt(j)) {
                case 'L':
                    d += 1;
                    break;
                case 'R':
                    d += 3;
                    break;
                case 'M':
                    switch (d % 4) {
                    case 0:
                        ++x;
                        break;
                    case 1:
                        ++y;
                        break;
                    case 2:
                        --x;
                        break;
                    case 3:
                        --y;
                        break;
                    }
                    break;
                }
            }
            out.printf("%d %d %c%n", x, y, "ENWS".charAt(d % 4));
        }
    }
}

File: src/main/kotlin/hm/binkley/rover/CStyleMain.kt
package hm.binkley.rover

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.lang.System.`in`
import java.lang.System.out
import java.util.stream.Collectors.toList

/**
 * `CStyleMain` is a very "C"-like approach to the Rover problem.
 *
 * See [Mars Rover in Python and Haskell](http://arunrocks.com/mars-rover-in-python-and-haskell/).
 */
object CStyleMain {
    @JvmStatic
    fun main(vararg args: String) {
        val lines = BufferedReader(InputStreamReader(`in`))
            .lines()
            .collect(toList())
        if (0 == lines.size % 2) throw IOException("Malformed input")

        // TODO: Skip grid size - what to do with it?
        var i = 1
        val l = lines.size
        while (i < l) {
            val co = lines[i]
            val coords = co.split(" ".toRegex(), 3).toTypedArray()
            var x = coords[0].toInt()
            var y = coords[1].toInt()
            var d = "ENWS".indexOf(coords[2][0])
            val mo = lines[++i]
            var j = 0
            val k = mo.length
            while (j < k) {
                when (mo[j]) {
                    'L' -> d += 1
                    'R' -> d += 3
                    'M' -> when (d % 4) {
                        0 -> ++x
                        1 -> ++y
                        2 -> --x
                        3 -> --y
                    }
                }
                ++j
            }
            out.printf("%d %d %c%n", x, y, "ENWS"[d % 4])
            ++i
        }
    }
}
